lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -0,0 +1,299 @@
+    return success();
+  }
+  return failure();
+}
+
+//===----------------------------------------------------------------------===//
+// Canonicalization pattern implementations
+//===----------------------------------------------------------------------===//
+
+LogicalResult MemWritePortEnableAndMaskCanonicalizer::matchAndRewrite(
+    MemoryWritePortOp op, PatternRewriter &rewriter) const {
+  auto defOp = cast<DefineOp>(symbolCache.getDefinition(op.getArcAttr()));
+  APInt enable;
+
+  if (op.getEnable() &&
+      mlir::matchPattern(
+          defOp.getBodyBlock().getTerminator()->getOperand(op.getEnableIdx()),
+          mlir::m_ConstantInt(&enable))) {
+    if (enable.isZero()) {
+      symbolCache.removeUser(defOp, op);
+      rewriter.eraseOp(op);
+      if (symbolCache.useEmpty(defOp)) {
+        symbolCache.removeDefinitionAndAllUsers(defOp);
+        rewriter.eraseOp(defOp);
+      }
+      return success();
+    }
+    if (enable.isAllOnes()) {
+      auto newName = names.newName(defOp.getName());
+      auto users = SmallVector<Operation *>(symbolCache.getUsers(defOp));
+      symbolCache.removeDefinitionAndAllUsers(defOp);
+
+      // Remove the enable attribute
+      rewriter.updateRootInPlace(op, [&]() {
+        op.setEnable(false);
+        op.setArc(newName);
+      });
+
+      auto newResultTypes = op.getArcResultTypes();
+
+      // Create a new arc that acts as replacement for other users
+      rewriter.setInsertionPoint(defOp);
+      auto newDefOp = rewriter.cloneWithoutRegions(defOp);
+      auto *block = rewriter.createBlock(
+          &newDefOp.getBody(), newDefOp.getBody().end(),
+          newDefOp.getArgumentTypes(),
+          SmallVector<Location>(newDefOp.getNumArguments(), defOp.getLoc()));
+      auto callOp = rewriter.create<CallOp>(newDefOp.getLoc(), newResultTypes,
+                                            newName, block->getArguments());
+      SmallVector<Value> results(callOp->getResults());
+      Value constTrue = rewriter.create<hw::ConstantOp>(
+          newDefOp.getLoc(), rewriter.getI1Type(), 1);
+      results.insert(results.begin() + op.getEnableIdx(), constTrue);
+      rewriter.create<OutputOp>(newDefOp.getLoc(), results);
+
+      // Remove the enable output from the current arc
+      auto *terminator = defOp.getBodyBlock().getTerminator();
+      rewriter.updateRootInPlace(
+          terminator, [&]() { terminator->eraseOperand(op.getEnableIdx()); });
+      rewriter.updateRootInPlace(defOp, [&]() {
+        defOp.setName(newName);
+        defOp.setFunctionType(
+            rewriter.getFunctionType(defOp.getArgumentTypes(), newResultTypes));
+      });
+
+      // Update symbol cache
+      symbolCache.addDefinition(defOp.getNameAttr(), defOp);
+      symbolCache.addDefinition(newDefOp.getNameAttr(), newDefOp);
+      symbolCache.addUser(defOp, callOp);
+      for (auto *user : users)
+        symbolCache.addUser(user == op ? defOp : newDefOp, user);
+      return success();
```

Comment:
Should this pattern have a cache somewhere that keeps track of the defOp-minus-enable for each defOp, such that you only create a def without enable for the first memory op, and subsequent ops would just use the cached def without enable?

Probably not a big thing in practice though :grin:

Reply:
Excellent point! I've added that.